Exercise 1: Cross-Entropy of two distributions (the pure concept)Goal: Understand cross-entropy as "surprise when reality is P but you expected Q."

The problemYou have a weather forecaster. The TRUE distribution of weather in Pune over a week is:
P = [0.7, 0.2, 0.1]    # [sunny, cloudy, rainy]

You're testing 3 different forecasters. Each gives their predicted distribution Q:
Forecaster A (perfect):   Q = [0.7, 0.2, 0.1]
Forecaster B (decent):      Q = [0.6, 0.3, 0.1]
Forecaster C (overconfident): Q = [0.95, 0.04, 0.01]
Forecaster D (terrible):    Q = [0.1, 0.1, 0.8]

Your task
Implement cross_entropy(P, Q) from scratch
Compute cross-entropy of each forecaster vs reality
Rank them from best to worst
Predict before computing — which one will have lowest cross-entropy? Highest?

In [15]:
import numpy as np

P = [0.7, 0.2, 0.1] # [sunny, cloudy, rainy]

In [16]:
def cross_entropy(P, Q):
    """
    Compute cross-entropy of distribution P (true) against Q (predicted).
    Both should be numpy arrays summing to 1.
    """
    P = np.array(P)
    Q = np.array(Q)

    # cross entropy: -sum(P(x) * log(Q(x)))
    return -np.sum(P * np.log2(Q))

In [17]:
forecaster = {
    "A (perfect)": np.array([0.7, 0.2, 0.1]),
    "B (decent)": np.array([0.6, 0.3, 0.1]),
    "C (overconfident)": np.array([0.95, 0.04, 0.01]),
    "D (terrible)": np.array([0.1, 0.1, 0.8])
}
print("Cross_entropy values (Lower the better):")
for name, Q in forecaster.items():
    ce = cross_entropy(P,Q)
    print(f" {name}, CE = {ce:.4f}")

Cross_entropy values (Lower the better):
 A (perfect), CE = 1.1568
 B (decent), CE = 1.1955
 C (overconfident), CE = 1.6450
 D (terrible), CE = 3.0219


In [18]:
# Compute entropy of P itself for reference
entropy_of_P = cross_entropy(P, P)
print(f"\nEntropy of P (the lower bound): {entropy_of_P:.4f}")


Entropy of P (the lower bound): 1.1568


##### Reflection questions

Why is forecaster A's cross-entropy NOT zero? (Hint: P itself has uncertainty — entropy of P is 1.157. Cross-entropy can't go below entropy of P. The minimum is reality's own entropy.)

Why is forecaster D so much worse than C? (D put 80% probability on rain, but it only rains 10% of the time. Massively wrong. Confidence in 
wrong direction = log explosion.)

What's the "perfect cross-entropy" score? (= entropy of P. You can't do better than knowing reality exactly.)